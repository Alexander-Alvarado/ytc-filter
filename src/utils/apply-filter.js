import RegexParser from 'regex-parser'
const regexCache = new Map()

export const applyFilter = (filter, msg) => {
  if (!filter) {
    return false
  }
  let caseSensitive
  let caseInsensitive
  let regex
  switch (filter.type) {
    case 'author':
      if (filter.value === '' || filter.value == null || msg.author === '') {
        return false
      }
      caseSensitive = filter.caseSensitive && msg.author === filter.author
      caseInsensitive = !filter.caseSensitive && msg.author.toLowerCase() === filter.value.toLowerCase()
      return caseSensitive || caseInsensitive

    case 'msgIncludes':
      if (filter.value === '' || filter.value == null || msg.message === '') {
        return false
      }
      caseSensitive = filter.caseSensitive && msg.message.includes(filter.msgIncludes)
      caseInsensitive = !filter.caseSensitive && msg.message.toLowerCase().includes(filter.value.toLowerCase())
      return caseSensitive || caseInsensitive

    case 'isMember':
      return Boolean(msg.member)

    case 'isModerator':
      return Boolean(msg.moderator)
    case 'isOwner':
      return Boolean(msg.owner)

    case 'isVerified':
      return Boolean(msg.verified)
    case 'isSuperChat':
      return msg.messageType === 'paid-message'
    case 'regex':
      try {
        regex = regexCache.get(filter.value)
        if (!regex) {
          regex = RegexParser(filter.value)
          regexCache.set(filter.value, regex)
        }
        return regex.test(msg.message)
      } catch (e) {
        console.log('[ytcFilter] Regex error:', e)
      }
  }
  return false
}
