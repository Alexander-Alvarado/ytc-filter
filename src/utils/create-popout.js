export const createYtcPopout = info => {
  chrome.windows.create(
    {
      url: `${chrome.runtime.getURL('popout.html')}?vid=${info.videoId}&vname=${encodeURIComponent(info.videoName)}&cid=${info.channelId}&cname=${info.channelName}`,
      type: 'popup',
    },
    function(window) {}
  )
}

export const createYoutubePopout = info => {
  chrome.windows.create({
    url: `https://www.youtube.com/live_chat?is_popout=1&v=${info.videoId}&ytc=1&vid=${info.videoId}&cid=${info.channelId}&cname=${info.channelName}`,
  })
}
