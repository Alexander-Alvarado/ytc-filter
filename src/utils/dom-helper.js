export const querySelectorAsync = (
    selector,
    interval = 100,
    timeout = 1000
  )=> {
    return new Promise((resolve) => {
      const expireTime = Date.now() + timeout
      const timer = setInterval(() => {
        const e = document.querySelector(selector)
        if (e || Date.now() > expireTime) {
          clearInterval(timer)
          resolve(e)
        }
      }, interval)
    })
  }
  
  export const getImageSourceAsync = (
    img,
    interval = 100,
    timeout = 1000
  ) => {
    return new Promise((resolve) => {
      const expireTime = Date.now() + timeout
      const timer = setInterval(() => {
        if (img.src || Date.now() > expireTime) {
          clearInterval(timer)
          resolve(img.src)
        }
      }, interval)
    })
  }
  
  export const waitImageLoaded = (
    img,
    interval = 100,
    timeout = 1000
  ) => {
    return new Promise((resolve) => {
      const expireTime = Date.now() + timeout
      const timer = setInterval(() => {
        if ((img.complete && img.naturalWidth) || Date.now() > expireTime) {
          clearInterval(timer)
          resolve(img.src)
        }
      }, interval)
    })
  }
  
  export const waitAllImagesLoaded = (
    element,
    interval = 100,
    timeout = 1000
  ) => {
    if (element instanceof HTMLImageElement) {
      return Promise.all([waitImageLoaded(element, interval, timeout)])
    }
    return Promise.all(
      Array.from(element.querySelectorAll('img')).map((img) => {
        return waitImageLoaded(img, interval, timeout)
      })
    )
  }