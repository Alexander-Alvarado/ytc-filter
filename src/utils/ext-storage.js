import copy from 'copy-to-clipboard'

class ExtStorage {
  get() {
    return new Promise((resolve, reject) => {
      try {
        chrome.storage.local.get('@@vwe-persistence', s => resolve(s['@@vwe-persistence'] || null))
      } catch (e) {
        console.error('[ytcFilter] extStorage get failed', e)
        reject(e)
      }
    })
  }

  async getYtcFilterVersion() {
    const storage = await this.get()
    return storage?.global?.version
  }

  async copyToClipboard() {
    copy(JSON.stringify(await this.get()))
  }
}

export const extStorage = new ExtStorage()
