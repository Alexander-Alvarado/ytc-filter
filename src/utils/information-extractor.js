export const getVideoId = () => {
  return new URLSearchParams(window.parent.top.location.search).get('v')
}

export const getChannelId = () => {
  return window.parent.top.document.querySelector('ytd-channel-name .yt-simple-endpoint').href.split('/').pop()
}