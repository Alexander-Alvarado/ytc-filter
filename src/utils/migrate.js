export const filterMigrate = filter => {
  if (filter.type) {
    return filter
  }
  const newFilter = {}
  for (const [key, value] of Object.entries(filter)) {
    if (key === 'caseSensitive') {
      newFilter.caseSensitive = value
    } else {
      newFilter.type = key
      newFilter.value = value
    }
  }
  return newFilter
}
