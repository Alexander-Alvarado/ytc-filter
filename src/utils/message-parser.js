import Color from 'color'
import { getImageSourceAsync } from '@/utils/dom-helper'

const getBackgroundColor = (el, opacity) => {
  try {
    const backgroundColor = getComputedStyle(el).backgroundColor
    const o = new Color(backgroundColor).object()
    return `rgba(${o.r}, ${o.g}, ${o.b}, ${opacity})`
  } catch (e) {
    // parse error by invalid background color
    return undefined
  }
}

const parseCommonElements = async el => {
  const id = el.id
  const author = el.querySelector('#author-name')?.textContent ?? undefined
  const authorType = el.getAttribute('author-type') ?? undefined
  const roles = {}
  switch (authorType) {
    case 'member':
      roles.member = true
      break
    case 'moderator':
      roles.moderator = true
      break
    case 'owner':
      roles.owner = true
      break
  }
  //const avatorImage = el.querySelector('#img')
  //const avatarUrl = (avatorImage && (await getImageSourceAsync(avatorImage))) ?? undefined
  const message = el.querySelector('#message')?.textContent ?? undefined
  for (const badge of el.querySelectorAll('yt-live-chat-author-badge-renderer')) {
    roles[badge.type] = true
  }
  const badgeUrl = el.querySelector('img.yt-live-chat-author-badge-renderer')?.src
  const verified = Boolean(el.querySelector('yt-live-chat-author-badge-renderer[type="verified"]'))
  return { id, message, author, authorType, /* avatarUrl,*/ badgeUrl, verified, ...roles }
}

const parseTextMessage = async el => {
  const params = await parseCommonElements(el)
  const html = el.querySelector('#message')?.innerHTML
  const timestamp = el.querySelector('#timestamp')?.innerHTML

  return {
    ...params,
    html,
    messageType: 'text-message',
    timestamp,
  }
}

const parsePaidMessage = async el => {
  const params = await parseCommonElements(el)

  const html = el.querySelector('#message')?.innerHTML
  const purchaseAmount = el.querySelector('#purchase-amount')?.textContent ?? undefined
  const card = el.querySelector('#card > #header')
  const backgroundColor = (card && getBackgroundColor(card, 0.8)) ?? undefined
  const timestamp = el.querySelector('#timestamp')?.innerHTML

  return {
    ...params,
    html,
    backgroundColor,
    purchaseAmount,
    messageType: 'paid-message',
    timestamp,
  }
}

const parsePaidSticker = async el => {
  const params = await parseCommonElements(el)

  const purchaseAmount = el.querySelector('#purchase-amount-chip')?.textContent ?? ''
  const card = el.querySelector('#card')
  const backgroundColor = (card && getBackgroundColor(card, 0.8)) ?? undefined
  const stickerImage = el.querySelector('#sticker > #img')
  const stickerUrl = (stickerImage && (await getImageSourceAsync(stickerImage))) ?? undefined

  return {
    ...params,
    stickerUrl,
    backgroundColor,
    purchaseAmount,
    messageType: 'paid-sticker',
  }
}

const parseMembershipItem = async el => {
  const params = await parseCommonElements(el)

  const detailText = el.querySelector('#header-subtext')?.textContent ?? undefined
  const header = el.querySelector('#card > #header')
  const backgroundColor = (header && getBackgroundColor(header, 0.8)) ?? undefined

  return {
    ...params,
    html: detailText,
    backgroundColor,
    messageType: 'membership-item',
  }
}

export const parse = async el => {
  const tagName = el.tagName.toLowerCase()
  switch (tagName) {
    case 'yt-live-chat-text-message-renderer':
      return await parseTextMessage(el)
    case 'yt-live-chat-paid-message-renderer':
      return await parsePaidMessage(el)
    case 'yt-live-chat-paid-sticker-renderer':
      return await parsePaidSticker(el)
    case 'yt-live-chat-membership-item-renderer':
      return await parseMembershipItem(el)
  }
}
