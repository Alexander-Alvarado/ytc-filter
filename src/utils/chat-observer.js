import { querySelectorAsync } from '@/utils/dom-helper'
import { parse } from '@/utils/message-parser'

export class ChatObserver {
  constructor() {
    this.listeners = []
  }

  async observe() {
    const items = await querySelectorAsync('#items.yt-live-chat-item-list-renderer')

    if (!items) {
      return
    }

    this.observer = new MutationObserver(mutations => {
      mutations.forEach(mutation => {
        const nodes = Array.from(mutation.addedNodes)
        nodes.forEach(node => {
          if (node instanceof HTMLElement) {
            this.proceed(node)
          }
        })
      })
    })

    this.observer.observe(items, { childList: true })

    // this.cleanupTimer = setInterval(() => {
    //   this.timelines = this.timelines.map((timelines) => {
    //     return timelines.filter((timeline) => {
    //       return timeline.didDisappear > Date.now()
    //     })
    //   })
    // }, 1000)
  }

  async proceed(element) {
    const message = await parse(element)
    if (!message) {
      return
    }

    for (const listener of this.listeners) {
      listener(message)
    }
  }

  clear() {
    this.observer?.disconnect()
  }
}
