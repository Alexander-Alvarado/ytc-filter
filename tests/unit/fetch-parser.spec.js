import { fetchParser } from '@/utils/fetch-parser'
import fetchAuthorless from './fixtures/fetch-authorless.txt'

describe('Fetch parser', () => {
  it('should parse a message even if there is no author', () => {
    const result = fetchParser(fetchAuthorless)
    expect(result).toMatchSnapshot()
  })
})
