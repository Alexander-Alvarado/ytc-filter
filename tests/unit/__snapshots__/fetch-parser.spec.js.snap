// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch parser should parse a message even if there is no author 1`] = `
Array [
  Object {
    "author": "すいれんP",
    "html": "トイレを中に仕込むのは難しそうだし離れが良いかもね",
    "id": "CjoKGkNPbjd1T2VvdnU0Q0ZiSWRyUVlkZHFNUDJnEhxDUENQOWNtZnZ1NENGUXlaY0FvZDViRU1DUTI3",
    "message": "トイレを中に仕込むのは難しそうだし離れが良いかもね",
    "timestamp": "10:31:21",
  },
  Object {
    "author": "パセリ",
    "html": "草",
    "id": "CjoKGkNQMzY1LWVvdnU0Q0ZTSUNyUVlkU1RNRzlREhxDSkRPb0wtWXZ1NENGUmZWZ2dvZFlwd09idzEx",
    "message": "草",
    "timestamp": "10:31:21",
  },
  Object {
    "author": "冴城躁庵",
    "html": "体に悪いから我慢しないで…",
    "id": "CjsKGkNORzE2LWVvdnU0Q0ZiZ1VyUVlkeF9zQUhREh1DS1Q4N2FHaXZ1NENGWU5QaXdvZE1oQUJ2Zy01Nw%3D%3D",
    "message": "体に悪いから我慢しないで…",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "C.Leo",
    "html": ";;",
    "id": "CjoKGkNJRHM3dWVvdnU0Q0ZiSWRyUVlkZHFNUDJnEhxDSjdDeWVPbHZ1NENGY1d1WXdZZHRNUUlmZzI3",
    "message": ";;",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "いなかのこう",
    "html": "そもそこまで我慢すな",
    "id": "CjkKGkNLUFg5ZWVvdnU0Q0ZTSUNyUVlkU1RNRzlREhtDTDNWNDVtbnZ1NENGUXRJaFFvZFNiNE5wQTA%3D",
    "message": "そもそこまで我慢すな",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "サバミソ",
    "badgeUrl": "https://yt3.ggpht.com/mC6-CXmx2uBJ8WbaOTgrdnEHzhKsyFxm8X8EWrPffYerOcFEyQEe6Fvp5W5nf-O6P_2kuojYXnc=s16-c-k",
    "html": "ｗｗ",
    "id": "CjoKGkNJUGwtZWVvdnU0Q0ZYMGVyUVlkVmxzR1BnEhxDTXZvaTRpWXZ1NENGVTlIaFFvZFlnQUw2ZzQ1",
    "member": true,
    "message": "ｗｗ",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "怪盗キット",
    "html": "和式トイレ",
    "id": "CjsKGkNLQ3k5LWVvdnU0Q0ZWWV9yUVlkRWtBSlRnEh1DSnk0aVlXb3Z1NENGYnRTaFFvZGt3b0VOdy0zMQ%3D%3D",
    "message": "和式トイレ",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "飛電_元hidekin",
    "badgeUrl": "https://yt3.ggpht.com/ccJj8YhVfO35nuxgczx6yPWe1dSDnTa8UCshIOuVwbGBW6O9XHGhHSFkGt8h30RjFfPi3kLsDg=s16-c-k",
    "html": "トイレの位置が何処にくるか、検討つけて建てないと",
    "id": "CjoKGkNQZmRfLWVvdnU0Q0Zjd0RyUVlkQkw4SHd3EhxDTWpHaU9DbnZ1NENGWkJGWUFvZHlPSU5idy0x",
    "member": true,
    "message": "トイレの位置が何処にくるか、検討つけて建てないと",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "Koay J-Shenn",
    "html": "F",
    "id": "CjoKGkNOYWVtLWlvdnU0Q0Zjd0RyUVlkQkw4SHd3EhxDUHlUcHFLaHZ1NENGY0ptS2dvZFpId09lQTIw",
    "message": "F",
    "timestamp": "10:31:22",
  },
  Object {
    "author": "とりう",
    "badgeUrl": "https://yt3.ggpht.com/Tmc31yIEGVmQvnnKmgNeBdm5SZ5j276adpMgCLVKxFRLj3htk5gVy0IiIqtJEIvh5wu-Zwtd=s16-c-k",
    "html": "炎症おこすぞｗ",
    "id": "CjoKGkNNLVpxLWlvdnU0Q0ZWWV9yUVlkRWtBSlRnEhxDSXVJMGZpZ3Z1NENGWGd4dHdBZGYtWUYwUTUy",
    "member": true,
    "message": "炎症おこすぞｗ",
    "timestamp": "10:31:23",
  },
  Object {
    "author": undefined,
    "badgeUrl": "https://yt3.ggpht.com/PQf1apgZon-YKiqmvHF-oQnjdMYthj5CAhlaH6jeSis5BqF9vp_lDC0OpTKN-nby0EmFF16Y-w=s16-c-k",
    "html": "膀胱炎…",
    "id": "CjsKGkNQM3FzLWlvdnU0Q0ZYMGVyUVlkVmxzR1BnEh1DS2picWJPWXZ1NENGWjlCOVFVZEdsNERXQS0xNA%3D%3D",
    "member": true,
    "message": "膀胱炎…",
    "timestamp": "10:31:23",
  },
  Object {
    "author": "リーンリーン",
    "html": "草",
    "id": "CjoKGkNKUE96dWlvdnU0Q0ZYY1hyUVlkZmswRS1REhxDSU9tMlltanZ1NENGWW45T0FZZE9MNENlZzE0",
    "message": "草",
    "timestamp": "10:31:23",
  },
  Object {
    "author": "American Gundam",
    "html": "so darn lovely",
    "id": "CjkKGkNNcmQwLWlvdnU0Q0ZiSWRyUVlkZHFNUDJnEhtDTUNrbU51b3Z1NENGV1VEVEFvZFRjc0hldzE%3D",
    "message": "so darn lovely",
    "timestamp": "10:31:23",
  },
  Object {
    "author": "K R B【かるび】ん",
    "badgeUrl": "https://yt3.ggpht.com/mC6-CXmx2uBJ8WbaOTgrdnEHzhKsyFxm8X8EWrPffYerOcFEyQEe6Fvp5W5nf-O6P_2kuojYXnc=s16-c-k",
    "html": "がまんした方がきもちいいぞ",
    "id": "CjoKGkNKSzgzdWlvdnU0Q0ZiSWRyUVlkZHFNUDJnEhxDSUhTdmJXZ3Z1NENGUlVJWUFvZG5OUUF2QTM4",
    "member": true,
    "message": "がまんした方がきもちいいぞ",
    "timestamp": "10:31:23",
  },
  Object {
    "author": "エスターク",
    "html": "頑張らなくていいんだよ",
    "id": "CjoKGkNJXzU5dWlvdnU0Q0ZUb3VyUVlkX2hnSnRREhxDTXF6N0tLb3Z1NENGVXlrV0FvZC1kSU5nUS0x",
    "message": "頑張らなくていいんだよ",
    "timestamp": "10:31:24",
  },
  Object {
    "author": "No Jam Jeong [Reine's Chat Food Provider]",
    "html": "F",
    "id": "CkUKGkNLbmZrZW1vdnU0Q0ZiZ1VyUVlkeF9zQUhREidDSi01dlptb3Z1NENGZkhaVEFJZHVyTUlMUTE2MTE4MjYyNzM0MjU%3D",
    "message": "F",
    "timestamp": "10:31:24",
  },
  Object {
    "author": "CHAKA24",
    "html": "RIP 🦃",
    "id": "CkUKGkNQUGRzT21vdnU0Q0ZYY1hyUVlkZmswRS1REidDTWFGNC1hb3Z1NENGVlNXd2dvZGxCa0k1dzE2MTE4MjYyODQ2NjU%3D",
    "message": "RIP 🦃",
    "timestamp": "10:31:25",
  },
  Object {
    "author": "名前のない猫",
    "badgeUrl": "https://yt3.ggpht.com/Tmc31yIEGVmQvnnKmgNeBdm5SZ5j276adpMgCLVKxFRLj3htk5gVy0IiIqtJEIvh5wu-Zwtd=s16-c-k",
    "html": "草",
    "id": "CjkKGkNPYnN0T21vdnU0Q0ZYY1hyUVlkZmswRS1REhtDSWJzaHNlWXZ1NENGUWVBWXdZZDRaVUFWUTk%3D",
    "member": true,
    "message": "草",
    "timestamp": "10:31:25",
  },
]
`;
